### PR TITLE
Add restart capability to theta nh runs

### PR DESCRIPTION
**Pull request by oksanaguba**
Adds restart to runs with cam_target=theta-l with hydrostatic_mode=.false. . Required reading/writing 2 dynamic variables on interfaces, w_i and phinh_i .

### CODE REVIEW

**Comment by oksanaguba** [comment]
Ran one 5 day run and five one day runs. Runs match:
```
[onguba@b724 run]$ pwd
/home/onguba/as/theta-5d-nh/run

mpiexec -n 1 /home/ccsm-data/tools/cprnc -m theta-5d-nh.cam.r.0001-01-06-00000.nc ~/as/theta-5d-restart-nh/run/theta-5d-restart-nh.cam.r.0001-01-06-00000.nc

mpiexec -n 1 /home/ccsm-data/tools/cprnc -m theta-5d-nh.cpl.r.0001-01-06-00000.nc ~/as/theta-5d-restart-nh/run/theta-5d-restart-nh.cpl.r.0001-01-06-00000.nc
```

Scripts are below.

**Comment by oksanaguba** [comment]
```
#which=acme-b3
#fol=theta-5d-nh

#${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
#cd $fol

#./xmlchange CAM_TARGET=preqx
./xmlchange CAM_TARGET=theta-l
./xmlchange JOB_WALLCLOCK_TIME=1:00:00
#on skybridge on 8 nodes=128 ranks 5 days is 40 min
#here 4 nodes = 4*36 =144 ranks, 50 min.
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0

#./case.setup --reset
#./case.setup

uf=user_nl_cam
#first empty line
echo "" > $uf

#add your defs here
echo "theta_hydrostatic_mode=.false." >> $uf
echo "tstep_type=7" >> $uf
echo "se_nsplit=20" >> $uf
echo "cubed_sphere_map=0" >> $uf

#./case.build --clean
./case.build
./xmlchange STOP_N=5 ; ./xmlchange STOP_OPTION=ndays ;
./case.submit
```

and

```#${which}/cime/scripts/create_newcase -case $fol -res ne4_ne4 -mach anvil -compiler intel -compset FC5AV1C-L
#cd $fol

#./xmlchange CAM_TARGET=preqx
./xmlchange CAM_TARGET=theta-l
./xmlchange JOB_WALLCLOCK_TIME=1:00:00
#on skybridge on 8 nodes=128 ranks 5 days is 40 min
#here 4 nodes = 4*36 =144 ranks, 50 min.
./xmlchange NTASKS=16
./xmlchange --file env_mach_pes.xml NTASKS_ESP=1
./xmlchange ROOTPE=0

#./case.setup --reset
#./case.setup

uf=user_nl_cam
#first empty line
echo "" > $uf

#add your defs here
echo "theta_hydrostatic_mode=.false." >> $uf
echo "tstep_type=7" >> $uf
echo "se_nsplit=20" >> $uf
echo "cubed_sphere_map=0" >> $uf

#./case.build --clean
./case.build
./xmlchange STOP_N=1 ; ./xmlchange STOP_OPTION=ndays ;
./xmlchange RESUBMIT=4;
./case.submit

#we only need the latest ID
id=`grep -o case.run:\[0-9\]\* CaseStatus | sed -e 's/.*://' | tail -1`
echo $id
cp user_nl_cam usercam."${id}"
```

**Comment by oksanaguba** [comment]
Do we want hydrostatic restart from a nonhydrostatic run? This PR does not allow it. Also, I am not sure whether I put in enough checks, I only repeated code in restart module, if there are checks in some other files, I am not aware of them (yet).

File: components/cam/src/dynamics/se/restart_dynamics.F90 [hunk 1]
```diff
@@ -389,20 +460,42 @@ subroutine read_restart_dynamics (File, dyn_in, dyn_out, NLFileName)
     call PIO_InitDecomp(pio_subsystem, pio_double, (/ncols,nlev/),ldof , iodesc3d)
     deallocate(ldof)
 
+#ifdef MODEL_THETA_L
+    if ( .not. theta_hydrostatic_mode ) then
+       allocate(var3dp(s2d*nlevp))
+    endif
+    var3dp = 0.0
+
```
Comment (amametjanov) [inline components/cam/src/dynamics/se/restart_dynamics.F90]: `var3dp` appears to be unallocated when `theta_hydrostatic_mode=.true.`:
```
  4:
  4: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
  4:
  4: Backtrace for this error:
  4: #0  0x2b94e6fc227f in ???
  4: #1  0xd3ffa6 in __restart_dynamics_MOD_read_restart_dynamics
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/components/cam/src/dynamics/se/restart_dynamics.F90:467
  4: #2  0x567c28 in __cam_restart_MOD_cam_read_restart
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/components/cam/src/control/cam_restart.F90:397
  4: #3  0x5185ca in __cam_comp_MOD_cam_init
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/components/cam/src/control/cam_comp.F90:167
  4: #4  0x510b62 in __atm_comp_mct_MOD_atm_init_mct
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/components/cam/src/cpl/atm_comp_mct.F90:314
  4: #5  0x433174 in __component_mod_MOD_component_init_cc
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/cime/src/drivers/mct/main/component_mod.F90:248
  4: #6  0x421bc7 in __cime_comp_mod_MOD_cime_init
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/cime/src/drivers/mct/main/cime_comp_mod.F90:1337
  4: #7  0x42f5b9 in cime_driver
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/cime/src/drivers/mct/main/cime_driver.F90:122
  4: #8  0x42f73a in main
  4:    at /lcrc/group/acme/azamat/repos/E3SM-bebop-master/cime/src/drivers/mct/main/cime_driver.F90:23
```
This is during restart run of `ERS.ne4_ne4.FC5AV1C-L.anvil-centos7_gnu.cam-thetahy_ftype2` on latest master.

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: Thanks! I will push a fix. Would it be caught with DEBUG?

Comment (amametjanov) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: gcc-8.2.0 is running into this in both debug and non-debug mode. I'm not sure if intel will catch this in debug mode. Nightlies are in non-debug mode.

Comment (amametjanov) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: Ok, out-of-curiosity, Intel v.17.0.4 in debug mode reports a different error, later in the restart run:
```
 91: forrtl: error (65): floating invalid
 91: Image              PC                Routine            Line        Source
 91: e3sm.exe           0000000009ACF132  Unknown               Unknown  Unknown
 91: libpthread-2.17.s  00002B657BA545D0  Unknown               Unknown  Unknown
 91: e3sm.exe           0000000003208753  prim_state_mod_mp         230  prim_state_mod.F90
 91: e3sm.exe           0000000002FC8D2D  prim_driver_base_         949  prim_driver_base.F90
 91: e3sm.exe           00000000041E9AFE  dyn_comp_mp_dyn_i         322  dyn_comp.F90
 91: e3sm.exe           0000000002CEB34D  restart_dynamics_         681  restart_dynamics.F90
 91: e3sm.exe           00000000009EF860  cam_restart_mp_ca         397  cam_restart.F90
 91: e3sm.exe           000000000084AFA5  cam_comp_mp_cam_i         167  cam_comp.F90
 91: e3sm.exe           00000000008164E0  atm_comp_mct_mp_a         312  atm_comp_mct.F90
 91: e3sm.exe           0000000000465390  component_mod_mp_         257  component_mod.F90
 91: e3sm.exe           000000000042A509  cime_comp_mod_mp_        1337  cime_comp_mod.F90
 91: e3sm.exe           000000000045C119  MAIN__                    122  cime_driver.F90
```
No errors in non-debug mode.

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: Line 230 in prim_state_mod is 
`       psmin_local(ie) = MINVAL(tmp(:,:,ie))`
where 
`       tmp(:,:,ie)=hvcoord%hyai(1)*hvcoord%ps0 + sum(elem(ie)%state%dp3d(:,:,:,n0),3)`
I would assume there are nans in dp3d from some other issue. How do you run? Is it still a restart test? I would like to try.

Comment (amametjanov) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: Same test. This was on Bebop `ERS_D.ne4_ne4.FC5AV1C-L.bebop_intel.cam-thetahy_ftype2`.

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: I am running on anvil right now. can I use bebop? I could ssh to it.

Comment (amametjanov) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: Yes, anyone with access to Anvil should be able to run on Bebop after `ssh bebop.lcrc.anl.gov`.

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: I see the same on anvil. I will debug it. Meanwhile, could we please have `srun -K` for anvil? my job got stuck.

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: @amametjanov  see https://github.com/E3SM-Project/E3SM/pull/3426 , finally debugged this . thanks!

Comment (amametjanov) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: That is referring to ERS_D.ne4_ne4.FC5AV1C-L.anvil_gnu.cam-thetahy_ftype2. I thought the recent error was about anvil+intel (not gnu).

Comment (oksanaguba) [reply components/cam/src/dynamics/se/restart_dynamics.F90]: GNU run was also crashing in printstate mode. I ran with intel just now, all is ok:
```
(base) [onguba@blueslogin4 as]$ ./cs.status.20200204_200805_z29jlq
20200204_200805_z29jlq: 1 test
  ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 (Overall: PASS) details:
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 CREATE_NEWCASE
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 XML
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 SETUP
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 SHAREDLIB_BUILD time=154
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 MODEL_BUILD time=203
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 SUBMIT
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 RUN time=708
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 COMPARE_base_rest
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 MEMLEAK
    PASS ERS_D.ne4_ne4.FC5AV1C-L.anvil_intel.cam-thetahy_ftype2 SHORT_TERM_ARCHIVER
```

**Pull request merged**